### ISSUE TITLE
Support for azurerm_automation_dsc_nodeconfiguration and "content_embedded"

### ISSUE THREAD

**Issue by robcamer** [terraform-providers/terraform-provider-azurerm#7404]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

Want automate Azure Automation DSC ingestion, compilation, and assignment to Virtual Machines.  It appears that DSC content must be embedded into the terraform script  as "content_embedded" which limits its usefulness.

1. What is the proper values to put in content_embedded as it appears to be the compiled DSC, aka .mof.

2. It would be MUCH better if the resource ID or a link to the .mof in Azure Automation be an option to the value for "content_embedded" for azurerm_automation_dsc_nodeconfiguration.

azurerm_automation_dsc_nodeconfiguration

azurerm_automation_dsc_nodeconfiguration

### Potential Terraform Configuration

```hcl
resource "azurerm_resource_group" "example" {
  name     = "resourceGroup1"
  location = "West Europe"
}

resource "azurerm_automation_account" "example" {
  name                = "account1"
  location            = azurerm_resource_group.example.location
  resource_group_name = azurerm_resource_group.example.name

  sku {
    name = "Basic"
  }
}

resource "azurerm_automation_dsc_configuration" "example" {
  name                    = "test"
  resource_group_name     = azurerm_resource_group.example.name
  automation_account_name = azurerm_automation_account.example.name
  location                = azurerm_resource_group.example.location
  content_embedded        = "configuration test {}"
}

resource "azurerm_automation_dsc_nodeconfiguration" "example" {
  name                    = "test.localhost"
  resource_group_name     = azurerm_resource_group.example.name
  automation_account_name = azurerm_automation_account.example.name
  depends_on              = [azurerm_automation_dsc_configuration.example]

  content_embedded = <<mofcontent
instance of MSFT_FileDirectoryConfiguration as $MSFT_FileDirectoryConfiguration1ref
{
  ResourceID = "[File]bla";
  Ensure = "Present";
  Contents = "bogus Content";
  DestinationPath = "c:\\bogus.txt";
  ModuleName = "PSDesiredStateConfiguration";
  SourceInfo = "::3::9::file";
  ModuleVersion = "1.0";
  ConfigurationName = "bla";
};
instance of OMI_ConfigurationDocument
{
  Version="2.0.0";
  MinimumCompatibleVersion = "1.0.0";
  CompatibleVersionAdditionalProperties= {"Omi_BaseResource:ConfigurationName"};
  Author="bogusAuthor";
  GenerationDate="06/15/2018 14:06:24";
  GenerationHost="bogusComputer";
  Name="test";
};
mofcontent

}
```

### References

* #0000

**Comment by neil-yechenwei** [comment]
Thanks for opening this issue. 
For above steps 3-4, I assume you are requesting a new resource “https://docs.microsoft.com/en-us/rest/api/automation/dsccompilationjob/create”. 
For step 5, I assume you have to use virtual machine extension to implement it. Below is the example.

```
resource "azurerm_virtual_machine_extension" "test" {
  name                             = "DSCForLinux"
  virtual_machine_id     = azurerm_virtual_machine.test.id
  publisher                       = "Microsoft.OSTCExtensions"
  type                                = "DSCForLinux"
  type_handler_version = "2.70"

  settings = <<SETTINGS
                  {
                    "ConfigurationMode": "ApplyAndMonitor",
                    "ConfigurationModeFrequencyMins": "15",
                      "ExtensionAction": "Register",
                      "NodeConfigurationName": "TestConfig.NotWebServer",
                     "RefreshFrequencyMins": "30",                                          
                      "RegistrationUrl": "${azurerm_automation_account.test.dsc_server_endpoint}",
                      "RegistrationKey": "${azurerm_automation_account.test.dsc_primary_access_key}",
                    "RebootNodeIfNeeded": "false",
                    "ActionAfterReboot": "ContinueConfiguration",
                    "AllowModuleOverwrite": "false"
                  }
SETTINGS
}
```

**Comment by neil-yechenwei** [comment]
After do more investigation for above steps 3-4 you described, I found [dsc complication job api spec](https://github.com/Azure/azure-rest-api-specs/blob/master/specification/automation/resource-manager/Microsoft.Automation/stable/2015-10-31/dscCompilationJob.json) doesn't support delete action. So I assume it's not appropriate to be as TF candidate.

**Comment by robcamer** [comment]
Hi There, I have been able to execute Terraform that uploads a fairly simple DSC, upload locally compiled MOF file, and apply the MOF configuration to a Node successfully.  Technically all of the DSC / Automation Terraform resources are working for a simple TSC.  

The issue is that I have DSCs that I cannot compile locally due to complexity.  This means that I cannot compile the DSC locally to generate the MOF and upload using **azurerm_automation_dsc_nodeconfiguration**.  Instead I have to use PowerShell Script like this to upload the DSC and then kick off compilation within the Azure Automation account in two steps:

Step 1 --> Import into Azure Automation: run (from your DSC working directory):
                                `Import-AzAutomationDscConfiguration -AutomationAccountName "automationAccount"-ResourceGroupName "automation_rg" -SourcePath ".\DSC.ps1" -Force -Published`

Step 2 --> Compile run (from the DSC working directory):
`$config = Import-PowerShellDataFile .\DSC.shared.psd1`
`Start-AzAutomationDscCompilationJob -ResourceGroupName 'AZ-DE-AFC-MADE-D-COR-RGP-01' -AutomationAccountName 'automationAccount' -ConfigurationName 'DSC' -ConfigurationData $config`

The above PowerShell will upload the DSC and compile it in Azure Automation, which works great for the more complex DSCs that are required for this project.   

**REQUEST:**
Add a "compile" bool option to azurerm_automation_dsc_configuration such that once the DSC is uploaded, compilation is kicked off.

For Tear down, Terraform can remove the compiled MOF file state changes in Azure Automation when it removes the uploaded DSC file.

**Comment by neil-yechenwei** [comment]
@robcamer, sorry. I assume terraform cannot add compile feature since dsc complication job doesn't support delete action, which means it's not a real terraform resource. So it's not appropriate to be as TF candidate.

**Comment by neil-yechenwei** [comment]
@robcamer, I assume it's not appropriate to be added into azurerm_automation_dsc_configuration as a property since terraform cannot know whether it has been compiled or it needs to be compiled since compile job api would create a new job id at each apply. So I assume terraform cannot handle multiple compile jobs in the same azurerm_automation_dsc_configuration resource. For instance, terraform cannot know whether it has been compiled at second apply after run the first apply with "compile=true", then terraform would trigger compile again in each apply. For this scenario, suggests to use terraform [provisioner](https://www.terraform.io/docs/provisioners/local-exec.html#interpreter-examples).

**Comment by robcamer** [comment]
Thanks Neil, I implemented with provisioner.  I deploy the DSC to Automation via azurerm_automation_dsc_configuration to have full lifecycle on the DSC.  So when the DSC is updated and needs to be deployed again the updated DSC is picked-up and applied by Terraform.   

Challenge with provisioner is that it will only run upon initial creation.  provisioner does not run when terraform apply results in the update lifecycle.  This means that  the DSC will be updated properly via azurerm_automation_dsc_configuration but the provisioner will not perform a compile in the terraform update lifecycle.  That's a significant issue.  To workaround that issue, I added a trigger to the provisioner that basically forces DSC compilation every time the script runs.  I would rather only have compile happen when azurerm_automation_dsc_configuration  is updated on terraform apply.  In fact, it is a feature that it re-compiles upon update. 

resource "null_resource" "compileDscADDS" {
  provisioner "local-exec" {
    command = "Powershell command to compile DSC"
    interpreter = ["pwsh", "-Command"]
  }
  triggers = {
    always_run = timestamp()
  }
  depends_on = [null_resource.azureSignInPWSH]
}

In Summary, It would be preferred that compilation automatically happen with compile = true on azurerm_automation_dsc_configuration then force compilation on every script run with provisioner if possible.  

This only leaves the remaining issue that the DSC is removed by terraform destroy.  If the DSC is removed then ideally the compiled MOF in Azure Automation would also be removed.  With provisioner I could trigger on destroy to handle that *I think* but I wanted to point it out as another consideration.

**Comment by neil-yechenwei** [comment]
Yes. I assume it's the better way to use [triggers](https://www.terraform.io/docs/provisioners/null_resource.html#example-usage) to re-run provisioner while updating.

**Comment by doherty100** [comment]
@robcamer can you whare the full code of your provisioner I have exactly the same problem. Thanks!
